### DISCUSSION TITLE
Importing created package fails with ImportError: DLL load failed (Fortran90)

### DISCUSSION THREAD

**Discussion by RubendeBruin** [mesonbuild/meson-python#595] [Q&A] [answered]
I'm trying to migrate an existing project to meson.
It involves creating a python package from a single fortran function.

The package is here: https://github.com/RubendeBruin/specpart

It builds successfully on github actions on windows-latest.  But when I download and test the created wheel it fails with `ImportError: DLL load failed while importing specpart: The specified module could not be found.`

Tried to solve this by enforcing a static build by including 
`project('specpart', 'c', default_options : ['default_library=static','default_library=static'])` in the [build](https://github.com/RubendeBruin/specpart/blob/main/meson.build) file but the error remains.

details (copied from [readme.md](https://github.com/RubendeBruin/specpart/blob/main/README.md):
- generated wrapper using f2py and the pre-existing specpart.pyf file
- created meson build file:
- set meson-python as build backend
- used `python -m build` to build the package
- used Github Actions to perform the actual build on windows
- workflow runs fine
- created wheel (in the artifacts)
- download the wheel and install it using pip
- FAILS WITH:
```
(work11) C:\data\python>pip install specpart-0.0.6-cp311-cp311-win_amd64.whl --force-reinstall
Processing c:\data\python\specpart-0.0.6-cp311-cp311-win_amd64.whl
Installing collected packages: specpart
  Attempting uninstall: specpart
    Found existing installation: specpart 0.0.6
    Uninstalling specpart-0.0.6:
      Successfully uninstalled specpart-0.0.6
Successfully installed specpart-0.0.6

(work11) C:\data\python>python
Python 3.11.3 (tags/v3.11.3:f3909b8, Apr  4 2023, 23:49:59) [MSC v.1934 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import specpart
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: DLL load failed while importing specpart: The specified module could not be found.
>>>

**Comment by rgommers** [comment, marked as answer]
Hi @RubendeBruin, thanks for the nice reproducer and README questions. My immediate two thoughts:
1. If you want to distribute a wheel (i.e., run on a different machine than it was built, like you are doing here), you should run `delvewheel` over it first in CI. That will vendor the `libgfortran.dll` runtime you need and make the binary portable.
   - On Windows specifically, this may even be needed on the same machine, because Windows is missing RPATH functionality and hence `pip install .` is usually broken if a DLL from outside the package is needed and not on the library search path.
   - On Linux/macOS, the same holds for distribution - there the relevant tools are `auditwheel`/`delocate`. (or maybe `repair-wheel` for all platforms, I think it works fine but have no real experience with it) 
2. Glancing at the SciPy `meson.build` files, the one difference is that it uses ` link_language: 'fortran'` to build the extensions using Fortran. If `auditwheel` alone doesn't fix things for you, this may help.

**Reply by rgommers** [reply]
Sorry, edited my reply - it's `delvewheel` for Windows.

**Reply by dnicolodi** [reply]
@rgommers It may be appropriate to put a note about this in the f2py documentation when building with Meson is discussed.

Is there not a way to link `libgfortran` statically? I guess it is like linking `libstdc++` statically, thus it may require some special treatment, but it should be possible.

**Reply by rgommers** [reply]
Yes, good point. We probably should also make it more prominent in our own docs how this work, because it's such a large footgun. We actually do mention `auditwheel` and co, but only on https://mesonbuild.com/meson-python/explanations/design-old.html?highlight=auditwheel (side note: that page is rendered after all)

> Is there not a way to link `libgfortran` statically?

IIRC it's both hard technically and a license concern (GCC runtime exception: https://stackoverflow.com/questions/10130143/gpl-lgpl-and-static-linking)

**Comment by RubendeBruin** [comment]
Hi @rgommers , thanks for the reply!

adding only the  `link_language: 'fortran'` did not do the trick.
Including `delvewheel` does work and includes the required .dll files in the wheel.

Now to figure out how to properly include it in the CI but that is another topic.
